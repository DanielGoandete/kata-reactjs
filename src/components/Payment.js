import React,{ useState } from 'react'
import {FaTimes} from 'react-icons/fa'

const Payment = ({nameOfBooks,onDelete}) => {

    const [getTotalValue , setTotalValue] = useState(0)
    const [getDiscountValue , setDiscountValue] = useState(0)
    const [showCheckout , setshowCheckout] = useState(false)
   
    
    const getCheckout = () =>{
       
        
        setshowCheckout(true);
        let elem
        let count 
        let discount 
        if(nameOfBooks.length > 0){
            for(let i = 0; i < nameOfBooks.length; i++){
                elem = nameOfBooks[i].idInsert
                count=0
                discount = 1
                for(let j = 0; j < nameOfBooks.length; j++){
                   
                    if(elem !== nameOfBooks[j].idInsert){
                        console.log('elem',elem)
                        console.log('nameOfBooks[j].idInsert',nameOfBooks[j].idInsert)
                        discount ++
                    }else{

                    }
                }
            }
        }
       
      //function that will show the prices on checkout  
      showPrice(discount)
    }


    const showPrice =(discount)=>{
        let pricePerID
        let total = 0 
        let noDiscont
        if(nameOfBooks.length > 0){
            //if there is books will set the values on the checkout
            noDiscont  = (nameOfBooks.length - discount)* 50 
            pricePerID = nameOfBooks.length * 50
        }else{
            //if there is no books on the basket then will set 0 on the checkout
            noDiscont = 0
            pricePerID = 0
            discount = 1
            setshowCheckout(false);
        }
        total = pricePerID - noDiscont 
        if(discount === 1){
             setTotalValue(total)
             setDiscountValue(0)
        }else if(discount === 2){
             discount = total * 0.05
             setTotalValue(total - discount + noDiscont)
             setDiscountValue(5)
        }else if(discount === 3){
             discount = total * 0.1
             setTotalValue(total - discount + noDiscont)
             setDiscountValue(10)
         }else if(discount === 4){
             discount = total * 0.2
             setTotalValue(total - discount + noDiscont)
             setDiscountValue(20)
         }else{
             discount = total * 0.25
             setTotalValue(total - discount + noDiscont)
             setDiscountValue(25)
         }
     }

     
    return (
        <div > 
            <table className="table table-dark">
              <thead>
                <tr>
                    <th  data-testid="basket" scope="col">Your Basket:</th>
                </tr>
              </thead>
              <tbody>
                { nameOfBooks?.map((arrValue,index)=> (
                    <tr key={index}>
                        <th scope="row">{arrValue.name} <FaTimes style={{color:'red', cursor:'pointer'}} onClick={ () => onDelete(arrValue.id)}></FaTimes></th>
                    </tr>
                ))}
              </tbody>
            </table>   
            <button style={{display:"none"}} className='btn btn-success buttonBuy' id="checkout" onClick={()=> getCheckout()}>checkout</button> 
            {showCheckout  ? 
                (<>
                    <p id="discount" >Discount: {getDiscountValue}%</p>
                    <p>Total to pay:{getTotalValue} â‚¬</p>
                </>)
                : ''
            }
        </div>
    )
}


export default Payment