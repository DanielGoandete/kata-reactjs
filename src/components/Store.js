import React,{ useState } from 'react'
import cleanCode from '../../src/assets/images/cleanCode.png'
import cleanCoder from '../../src/assets/images/cleanCoder.png'
import cleanArchitecture from '../../src/assets/images/cleanArchitecture.jpeg'
import tdd from '../../src/assets/images/tdd.jpeg'
import refactoring from '../../src/assets/images/refactoring.jpeg'
import Payment from './Payment'

const Store = () => {

    const [getprices , setprices] = useState([])
    const [getNames , setNames] = useState([])
    const [getNameAndId , setNameAndId] = useState([])
    const [getCheckoutNumber , setCheckoutNumber] = useState(0)
    const listIds = []
    const listNames =[]

    let booksData = [
        {id:Math.floor(Math.random() * 10000) +1, idInsert:0, name:'Clean Code' ,imgSrc:cleanCode, alt:"A HandBook of Agile Software Craftmanship"},
        {id:Math.floor(Math.random() * 10000) +1, idInsert:1, name:'Clean Coder' ,imgSrc:cleanCoder,alt:"A Code of Conduct Professional Programmers"},
        {id:Math.floor(Math.random() * 10000) +1, idInsert:2, name:'Clean Architecture' ,imgSrc:cleanArchitecture,alt:"A Craftman's Guide to Software Structure and Design"},
        {id:Math.floor(Math.random() * 10000) +1, idInsert:3, name:'TDD' ,imgSrc:tdd,alt:"Test-Driven Development"},
        {id:Math.floor(Math.random() * 10000) +1, idInsert:4, name:'Refactoring' ,imgSrc:refactoring,alt:"Working Effectively"}
    ];
  
    const getBookById =(id,name,idInsert)=>{
        
        listIds.push([id])
        listNames.push([name])
        getprices.push(listIds)
        setprices([...getprices])
        getNames.push(listNames)
        setNames([...getNames])
        const newNameId = {id , name,idInsert}
        getNameAndId.push(newNameId)
        setNameAndId([...getNameAndId])
        document.getElementById("checkout").click();
        setCheckoutNumber(getNameAndId?.length)
    }

 
    const deleteBook = (id)=>{

        setNameAndId(getNameAndId.filter((nameAndId)=> nameAndId.id !==id))
        for(let listOFPrices of getprices){
            
           // will remove the elem from the list 
             if(parseInt(listOFPrices) === id){
                let array  = getNameAndId.indexOf(id)
                getNameAndId.splice(array)
            }
        }
        document.getElementById("checkout").click();
        setCheckoutNumber(getNameAndId?.length)
    }

    //redirect to payment
    const toPay = ()=>{
        document.getElementById('payment').scrollIntoView();
    }
    //redirect to store
    const toBuy = ()=>{
        document.getElementById('toBuy').scrollIntoView();
     }


    return (
        <div id='toBuy'>
            <div  className="screen">
                <div className="row ">
                {
                    booksData?.map((bookDetail,index)=>{
                        return(
                            <div className="col"  key={index}>
                            <img src={bookDetail.imgSrc} alt={bookDetail.alt} data-testid="img"  width="90%" height="100%" />  
                            <button className='btn btn-success buttonBuy' data-testid="buy" onClick={()=> getBookById(bookDetail.id,bookDetail.name,bookDetail.idInsert)}>Buy</button>     
                            </div>
                        )
                })}
                </div>
                <p data-testid="basketItens" style={{display:"none"}}>{getCheckoutNumber}</p> 
                <button className='btn btn-success buttoCheckout' onClick={toPay}>{getCheckoutNumber} item(s): click in here to go Checkout</button>
            </div>
            <div className='section screen center' id="payment">
                <Payment nameOfBooks={getNameAndId} onDelete={deleteBook}></Payment>
                <button className='btn btn-success buttoCheckout' onClick={toBuy}>Back to Store</button>
            </div> 
        </div>
        
    )
}
export default Store