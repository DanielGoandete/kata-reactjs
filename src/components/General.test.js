import React from 'react'
import {render} from "@testing-library/react";
import Store from './Store'
import Payment from './Payment'


describe("initial renders", ()=>{

    it("check all  buttons buy ",()=>{
        const {getAllByTestId} = render(<Store/>)
        const button = getAllByTestId("buy")
        expect(button).toBeTruthy()
    })

    it("check all images ",()=>{
        const {getAllByTestId} = render(<Store/>)
        const img = getAllByTestId("img")
        expect(img).toBeTruthy()
    })

    it("check if basket Itens starts with 0 ",()=>{
        const {getByTestId} = render(<Store/>)
        const discount = getByTestId("basketItens")
        expect(discount.textContent).toBe("0")
    })

    it("check if my basket name is in there",()=>{
        const {getByTestId} = render(<Payment/>)
        const elem = getByTestId("basket")
        expect(elem.textContent).toBe("Your Basket:")
    })

    it("Snapshot Image test",()=>{
        expect(Store).toMatchSnapshot();
   })

   it("Snapshot Payment test",()=>{
        expect(Payment).toMatchSnapshot();
   })
    

})