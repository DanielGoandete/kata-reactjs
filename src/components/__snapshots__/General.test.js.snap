// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial renders Snapshot Image test 1`] = `[Function]`;

exports[`initial renders Snapshot Payment test 1`] = `[Function]`;
